### PR TITLE
ES6 Proxy

### PR DESCRIPTION
**Pull request by amir-arad**
naive solution proposal for #652

uses shallow observable map internally, because it fits the data structure's dynamic nature.

done as part of a #goodnessSquad hackathon

### CODE REVIEW

**Comment by amir-arad** [comment]
@idoros

File: src/api/observable.ts [hunk 1]
```diff
@@ -116,6 +116,35 @@ export class IObservableFactories {
...
+				}
+				return target.get(property);
+			},
+			set(target, property:string, value, receiver){
+				if (property === '$mobx'){
+					return false;}
+				target.set(property, value);
+				return true;
+			},
+			ownKeys: function(target) {
```
Comment (solkimicreb) [inline src/api/observable.ts]: This is used by `for in` loops internally too, so they work out of the box with expando props.

File: src/api/observable.ts [hunk 2]
```diff
@@ -116,6 +116,35 @@ export class IObservableFactories {
...
+				return target.keys();
+			},
+			getOwnPropertyDescriptor: function(target, property:string) {
```
Comment (solkimicreb) [inline src/api/observable.ts]: Is this necessary? Why not just leave the default property descriptor? If it is necessary shouldn't it include `writable: true` and `value: target.get(prop)`?

File: src/api/observable.ts [hunk 3]
```diff
@@ -116,6 +116,35 @@ export class IObservableFactories {
 		return res as any;
 	}
 
+	dynamic<T>(props: T, name?: string): T & IObservableObject {
```
Comment (solkimicreb) [inline src/api/observable.ts]: Maybe check of the props object is already wrapped by a dynamic observable Proxy. Accidental double wrapping could cause some nasty bugs.

File: src/api/observable.ts [hunk 4]
```diff
@@ -116,6 +116,35 @@ export class IObservableFactories {
...
+		if (arguments.length > 2)
+			incorrectlyUsedAsDecorator("object");
+		invariant(typeof Proxy === 'function', "dynamic objects are not supported in this environment");
+		const internalMap = new ObservableMap(props, referenceEnhancer, name);
+		const result = new Proxy(internalMap, {
```
Comment (solkimicreb) [inline src/api/observable.ts]: I think you could also add a `deleteProperty` and maybe a `has` trap for the sake of completeness.

**Comment by solkimicreb** [comment]
I don't know much about MobX, but I added a few general tips. Nice and clean approach in my opinion 🙂

File: src/api/observable.ts (hunk 2 above)
Comment (amir-arad) [inline src/api/observable.ts]: before we added `getOwnPropertyDescriptor`, Object.keys(foo) always returned an empty array, no matter what `ownKeys` actually returned.

Comment (amir-arad) [inline src/api/observable.ts]: AFAIK he rest of the fields have default fallbacks that fits our use case. `configurable` specifically breaks the Proxy when it does not match the internal property's `configurable` value, that is why we added it as well.

File: src/api/observable.ts (hunk 3 above)
Comment (amir-arad) [inline src/api/observable.ts]: I didn't think about that. thanks!

**Comment by amir-arad** [comment]
TODOs:
 - [ ] test and avoid double-wrap on already boxed argument
 - [ ] implement other proxy traps (`has`, `delete`)
 - [ ] test and support spy and observe events
Thanks for the feedback @solkimicreb !

**Comment by solkimicreb** [comment]
Deep wrapping will be a tricky one. It won't be enough to traverse the structure at start and wrap everything into an observable, because of the potential object valued expando properties.

**Comment by mweststrate** [comment]
Proxies are no on their way! So closing this PR in favor #1380, but thanks for the setup @amir-arad! Will borrow some stuff from this. Sad it took a year, but finally browser support is up to par :)  (although, IE, RN android, sigh)

**Pull request closed without merging**